mm2src/coins/z_coin/storage.rs

```diff
@@ -1,5 +1,215 @@
+use crate::z_coin::ValidateBlocksError;
+use crate::z_coin::ZcoinConsensusParams;
+
 pub mod blockdb;
 pub use blockdb::*;
 
 pub mod walletdb;
 pub use walletdb::*;
+
+#[cfg(target_arch = "wasm32")]
+use walletdb::wallet_idb_storage::DataConnStmtCacheWasm;
+#[cfg(debug_assertions)]
+use zcash_client_backend::data_api::error::Error;
+use zcash_client_backend::proto::compact_formats::CompactBlock;
+#[cfg(not(target_arch = "wasm32"))]
+use zcash_client_sqlite::with_async::DataConnStmtCacheAsync;
+use zcash_primitives::block::BlockHash;
+use zcash_primitives::consensus::BlockHeight;
+
+cfg_native!(
+    use zcash_client_backend::data_api::PrunedBlock;
+    use zcash_client_backend::wallet::{AccountId, WalletTx};
+    use zcash_client_backend::welding_rig::scan_block;
+    use zcash_extras::{WalletRead, WalletWrite};
+    use zcash_primitives::merkle_tree::CommitmentTree;
+    use zcash_primitives::sapling::Nullifier;
+    use zcash_primitives::zip32::ExtendedFullViewingKey;
+);
+
+#[derive(Clone)]
+pub struct DataConnStmtCacheWrapper {
+    #[cfg(not(target_arch = "wasm32"))]
+    pub cache: DataConnStmtCacheAsync<ZcoinConsensusParams>,
+    #[cfg(target_arch = "wasm32")]
+    pub cache: DataConnStmtCacheWasm,
+}
+
+impl DataConnStmtCacheWrapper {
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn new(cache: DataConnStmtCacheAsync<ZcoinConsensusParams>) -> Self { Self { cache } }
+    #[cfg(target_arch = "wasm32")]
+    pub fn new(cache: DataConnStmtCacheWasm) -> Self { Self { cache } }
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn inner(&self) -> DataConnStmtCacheAsync<ZcoinConsensusParams> { self.clone().cache }
+    #[cfg(target_arch = "wasm32")]
+    pub fn inner(&self) -> DataConnStmtCacheWasm { self.clone().cache }
+}
+
+#[allow(unused)]
+pub struct CompactBlockRow {
+    pub(crate) height: BlockHeight,
+    pub(crate) data: Vec<u8>,
+}
+
+#[derive(Clone)]
+pub enum BlockProcessingMode {
+    Validate,
+    Scan(DataConnStmtCacheWrapper),
+}
+
+/// Checks that the scanned blocks in the data database, when combined with the recent
+/// `CompactBlock`s in the cache database, form a valid chain.
+///
+/// This function is built on the core assumption that the information provided in the
+/// cache database is more likely to be accurate than the previously-scanned information.
+/// This follows from the design (and trust) assumption that the `lightwalletd` server
+/// provides accurate block information as of the time it was requested.
+///
+pub async fn validate_chain(
+    block: CompactBlock,
+    prev_height: &mut BlockHeight,
```

Comment:
Q: Is there a way to reach `prev_height` from `block` ?

Reply:
nop actually.